bob-access/src/extractor.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{credentials::{Credentials, CredentialsBuilder}, error::Error};
+use axum::extract::RequestParts;
+use http::Request;
+use tonic::{transport::server::TcpConnectInfo, Request as TonicRequest};
+use super::CredentialsType;
+
+pub trait Extractor {
+    fn get_header(&self, header: &str) -> Result<Option<&str>, Error>;
+    fn get_extension<T: Send + Sync + 'static>(&self) -> Option<&T>;
+}
+
+pub trait ExtractorExt {
+    fn extract(&self, cred_type: CredentialsType) -> Result<Credentials, Error>;
+    fn extract_basic(&self) -> Result<Credentials, Error>;
+    fn extract_token(&self) -> Result<Credentials, Error>;
+    fn extract_internode(&self) -> Result<Credentials, Error>;
+}
+
+fn prepare_builder<T: Extractor>(slf: &T) -> Result<CredentialsBuilder, Error> {
+    let addr = slf
+        .get_extension::<TcpConnectInfo>()
+        .and_then(|ext| ext.remote_addr());
+    let mut builder = Credentials::builder();
+    builder.with_address(addr);
+    Ok(builder)
+}
+
+impl<T: Extractor> ExtractorExt for T {
+    fn extract(&self, cred_type: CredentialsType) -> Result<Credentials, Error> {
+        match cred_type {
+            CredentialsType::None => {
+                return Ok(Credentials::default());
+            },
+            CredentialsType::Basic => {
+                return self.extract_basic().or_else(|e| {
+                    self.extract_internode().map_err(|_| e)
```

Comment:
InterNode credentials is required for `Basic` authenticator, so its extraction can be integrated into `extract_basic`. Same for `extract_token`